File: scripts/verify_block_coverage.py
#!/usr/bin/env python3
"""Verify block coverage for BIP-110 chain scan.

Checks:
- chunks.meta total_blocks vs scan_merged.json blocks_scanned
- Era block counts sum to total
- Batch block counts sum to total
- Chunk files exist
- Optional: known block hashes (genesis) if index readable
"""
import json
import os
import sys
from pathlib import Path

def main():
    chunks_dir = Path(os.environ.get("BLOCK_CACHE_DIR", "/run/media/acolyte/Extra/blockchain"))
    scan_path = Path("bip110_results/scan_merged.json")
    if not scan_path.exists():
        scan_path = Path("blvm-bench/bip110_results/scan_merged.json")
    if not scan_path.exists():
        scan_path = Path(__file__).parent.parent / "bip110_results" / "scan_merged.json"

    print("🔍 Block Coverage Verification")
    print(f"   Chunks directory: {chunks_dir}")
    print()

    errors = []
    warnings = []

    # 1. Load chunks.meta
    meta_path = chunks_dir / "chunks.meta"
    if not meta_path.exists():
        errors.append(f"chunks.meta not found at {meta_path}")
        total_blocks = None
        num_chunks = None
    else:
        meta = {}
        for line in meta_path.read_text().splitlines():
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                meta[k.strip()] = v.strip()
        total_blocks = int(meta.get("total_blocks", 0))
        num_chunks = int(meta.get("num_chunks", 0))
        blocks_per_chunk = int(meta.get("blocks_per_chunk", 0))
        print("📋 Metadata (chunks.meta):")
        print(f"   total_blocks: {total_blocks}")
        print(f"   num_chunks: {num_chunks}")
        print(f"   blocks_per_chunk: {blocks_per_chunk}")
        print()

    # 2. Check chunk files
    if num_chunks is not None:
        missing = []
        for i in range(num_chunks):
            if not (chunks_dir / f"chunk_{i}.bin.zst").exists():
                missing.append(i)
        if missing:
            errors.append(f"Missing chunk files: {missing}")
        else:
            print(f"✅ All {num_chunks} chunk files present")
        print()

    # 3. Load scan_merged.json
    if not scan_path.exists():
        warnings.append(f"scan_merged.json not found at {scan_path}")
        blocks_scanned = None
        era_blocks = None
        batch_blocks = None
    else:
        with open(scan_path) as f:
            scan = json.load(f)
        blocks_scanned = scan.get("blocks_scanned")
        print("📋 Scan results (scan_merged.json):")
        print(f"   blocks_scanned: {blocks_scanned}")
        print()

        # 4. Cross-check blocks_scanned vs total_blocks
        if total_blocks is not None and blocks_scanned is not None:
            if blocks_scanned == total_blocks:
                print("✅ blocks_scanned matches chunks.meta total_blocks")
            else:
                errors.append(f"Mismatch: scan has {blocks_scanned}, metadata has {total_blocks}")
        print()

        # 5. Era block sum
        era_blocks = 0
        for key in ["era_pre_segwit", "era_segwit", "era_taproot", "era_inscriptions"]:
            e = scan.get(key, {})
            if isinstance(e, dict):
                era_blocks += e.get("blocks", 0)
        if blocks_scanned is not None and era_blocks == blocks_scanned:
            print("✅ Era blocks sum matches blocks_scanned")
        elif era_blocks:
            print(f"   Era sum: {era_blocks} (expected {blocks_scanned})")
            if era_blocks != blocks_scanned:
                warnings.append(f"Era sum {era_blocks} != blocks_scanned {blocks_scanned}")
        print()

        # 6. Batch consistency (from individual batch files if available)
        batch_dir = scan_path.parent
        batch_sum = 0
        batch_files = [
            ("scan_0_400k.json", 400001),
            ("scan_400k_600k.json", 200000),
            ("scan_600k_900k.json", 312722),
        ]
        for fname, expected in batch_files:
            p = batch_dir / fname
            if p.exists():
                with open(p) as f:
                    b = json.load(f)
                batch_sum += b.get("blocks_scanned", 0)
        if batch_sum and batch_sum == blocks_scanned:
            print("✅ Batch blocks sum matches blocks_scanned")
        elif batch_sum:
            print(f"   Batch sum: {batch_sum} (expected {blocks_scanned})")
            if batch_sum != blocks_scanned:
                warnings.append(f"Batch sum {batch_sum} != blocks_scanned {blocks_scanned}")
        print()

        # 7. Post-inscriptions era check
        pi_blocks = scan.get("blocks_post_inscriptions")
        era_inscriptions = scan.get("era_inscriptions", {})
        if isinstance(era_inscriptions, dict):
            ei_blocks = era_inscriptions.get("blocks", 0)
            if pi_blocks is not None and ei_blocks == pi_blocks:
                print("✅ blocks_post_inscriptions matches era_inscriptions.blocks")
            elif pi_blocks and ei_blocks:
                print(f"   Post-inscriptions: {pi_blocks}, era_inscriptions: {ei_blocks}")
        print()

    # Summary
    if errors:
        print("❌ ERRORS:")
        for e in errors:
            print(f"   {e}")
        sys.exit(1)
    if warnings:
        print("⚠️  Warnings:")
        for w in warnings:
            print(f"   {w}")
    print()
    print("✅ Block coverage verification PASSED")
    if total_blocks:
        print(f"   Full chain coverage: heights 0..{total_blocks-1} ({total_blocks} blocks)")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: src/bin/verify_block_coverage.rs
//! Verify that block chunks provide complete coverage of the blockchain.
//!
//! Checks:
//! - chunks.meta total_blocks vs index entries
//! - All heights 0..total_blocks-1 present (no gaps)
//! - Chunk files exist for all referenced chunks
//! - Optional: cross-check with scan_merged.json blocks_scanned

use anyhow::{Context, Result};
use std::collections::BTreeSet;
use std::path::PathBuf;

fn main() -> Result<()> {
    let chunks_dir = std::env::var("BLOCK_CACHE_DIR")
        .unwrap_or_else(|_| "/run/media/acolyte/Extra/blockchain".to_string());
    let chunks_dir = PathBuf::from(&chunks_dir);

    println!("🔍 Block Coverage Verification");
    println!("   Chunks directory: {}", chunks_dir.display());
    println!();

    // 1. Load metadata
    let metadata = blvm_bench::chunked_cache::load_chunk_metadata(&chunks_dir)?
        .ok_or_else(|| anyhow::anyhow!("No chunk metadata found (chunks.meta)"))?;

    println!("📋 Metadata (chunks.meta):");
    println!("   total_blocks: {}", metadata.total_blocks);
    println!("   num_chunks: {}", metadata.num_chunks);
    println!("   blocks_per_chunk: {}", metadata.blocks_per_chunk);
    println!();

    // 2. Verify chunk files exist
    let mut missing_chunks = Vec::new();
    for i in 0..metadata.num_chunks {
        let chunk_file = chunks_dir.join(format!("chunk_{}.bin.zst", i));
        if !chunk_file.exists() {
            missing_chunks.push(i);
        }
    }
    if !missing_chunks.is_empty() {
        anyhow::bail!(
            "Missing chunk files: {:?}. Expected chunks 0..{}",
            missing_chunks,
            metadata.num_chunks - 1
        );
    }
    println!(
        "✅ All {} chunk files present (chunk_0..chunk_{})",
        metadata.num_chunks,
        metadata.num_chunks - 1
    );
    println!();

    // 3. Load block index
    let index_path = chunks_dir.join("chunks.index");
    if !index_path.exists() {
        anyhow::bail!(
            "Block index not found at {}. Run build_block_index first.",
            index_path.display()
        );
    }

    let index_data = std::fs::read(&index_path)
        .with_context(|| format!("Failed to read index: {}", index_path.display()))?;
    let index: blvm_bench::chunk_index::BlockIndex =
        bincode::deserialize(&index_data).context("Failed to deserialize block index")?;

    println!("📋 Block index (chunks.index):");
    println!("   entries: {}", index.len());
    println!();

    // 4. Check coverage: index.len() vs total_blocks
    let expected = metadata.total_blocks as usize;
    if index.len() != expected {
        anyhow::bail!(
            "Index entry count mismatch: index has {} entries, metadata expects {}",
            index.len(),
            expected
        );
    }
    println!(
        "✅ Index entry count matches metadata ({} blocks)",
        expected
    );

    // 5. Check for gaps: every height 0..total_blocks-1 must exist
    let heights: BTreeSet<u64> = index.keys().copied().collect();
    let mut gaps = Vec::new();
    for h in 0..metadata.total_blocks {
        if !heights.contains(&h) {
            gaps.push(h);
            if gaps.len() > 20 {
                gaps.push(u64::MAX); // sentinel for "and more"
                break;
            }
        }
    }
    if gaps.last() == Some(&u64::MAX) {
        gaps.pop();
        anyhow::bail!(
            "Missing heights in index (gaps): {:?} ... and more ({} total missing)",
            gaps,
            (0..metadata.total_blocks)
                .filter(|h| !heights.contains(h))
                .count()
        );
    }
    if !gaps.is_empty() {
        anyhow::bail!("Missing heights in index (gaps): {:?}", gaps);
    }
    println!(
        "✅ No gaps: all heights 0..{} present",
        metadata.total_blocks - 1
    );

    // 6. Check for blocks in chunk 999 (missing)
    let missing_chunk_999: Vec<u64> = index
        .iter()
        .filter(|(_, e)| e.chunk_number == 999)
        .map(|(h, _)| *h)
        .collect();
    if !missing_chunk_999.is_empty() {
        println!(
            "⚠️  {} blocks in chunk_missing (chunk 999): {:?}...",
            missing_chunk_999.len(),
            &missing_chunk_999[..missing_chunk_999.len().min(10)]
        );
    } else {
        println!("✅ No blocks in chunk_missing (all in regular chunks)");
    }

    // 7. Verify chunk numbers are valid
    let invalid_chunks: Vec<(u64, usize)> = index
        .iter()
        .filter(|(_, e)| e.chunk_number >= metadata.num_chunks && e.chunk_number != 999)
        .map(|(h, e)| (*h, e.chunk_number))
        .take(10)
        .collect();
    if !invalid_chunks.is_empty() {
        anyhow::bail!(
            "Invalid chunk numbers (>= num_chunks): {:?}",
            invalid_chunks
        );
    }
    println!(
        "✅ All chunk references valid (0..{} or 999)",
        metadata.num_chunks - 1
    );
    println!();

    // 8. Optional: cross-check with scan_merged.json
    let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let alt_paths = [
        cwd.join("bip110_results").join("scan_merged.json"),
        cwd.join("blvm-bench")
            .join("bip110_results")
            .join("scan_merged.json"),
        PathBuf::from("blvm-bench/bip110_results/scan_merged.json"),
    ];
    let mut scan_json = None;
    for p in alt_paths.into_iter() {
        if p.exists() {
            if let Ok(data) = std::fs::read_to_string(&p) {
                if let Ok(v) = serde_json::from_str::<serde_json::Value>(&data) {
                    scan_json = Some((p, v));
                    break;
                }
            }
        }
    }
    if let Some((path, v)) = scan_json {
        if let Some(n) = v.get("blocks_scanned").and_then(|x| x.as_u64()) {
            println!("📋 Cross-check with scan_merged.json:");
            println!("   Path: {}", path.display());
            println!("   blocks_scanned: {}", n);
            if n == metadata.total_blocks {
                println!("✅ blocks_scanned matches total_blocks ({})", n);
            } else {
                println!(
                    "⚠️  Mismatch: scan has {}, metadata has {}",
                    n, metadata.total_blocks
                );
            }
        }
    } else {
        println!("ℹ️  scan_merged.json not found (optional cross-check skipped)");
    }

    println!();
    println!("✅ Block coverage verification PASSED");
    println!(
        "   Full chain coverage: heights 0..{} ({} blocks)",
        metadata.total_blocks - 1,
        metadata.total_blocks
    );

    Ok(())
}
